### PR TITLE
Sylvanc parser

### PR DESCRIPTION
**Pull request by sylvanc**
This PR adds a typed-AST parser.

### CODE REVIEW

**Review by rengolin — APPROVED** [review]
Hi Sylvan,

I haven't read everything yet, but it seems similar to the other front-end, but with a nicer AST to build from, so cleaner.

A few (non-blocking) comments...
* Some files have great comments, others, none. It would be nice if all header files had a general comment explaining why it exists in addition to the method explainers. This would help us understand the design decisions, for example, behind the `wellformed` pass, the graph traversal, etc. The ast, the parser and the lexer are very lean on comments.
* The filesystem implementation probably works fine on all three OSs we care about, but filesystem implementation is often subtle and mostly broken, so I'd be more comfortable (long-term) with an established implementation. They're all broken, too, but hopefully in predictable ways. LLVM has one that is confusing, but kinda works in all cases we care about, and we already depend on LLVM...
* I don't understand what we gain by using `wellformed << ast` instead of just having a method called `verify`. Not that it's a problem, but it seems to add boilerplate to the verification and I can't see what's the gain.

Overall, I don't mind this going in as is. It's disconnected from the other parts and it's better if we develop it in tree than on a private branch on your private repo. 

I'm approving the PR but maybe check with others to make sure no one else has any comment. I'll also try to finish reading the whole set, but this shouldn't change how I feel about it going in early rather than late. I'd rather fix any remaining issue in tree.

File: src/parser/parser.cc [hunk 1]
```diff
@@ -0,0 +1,2496 @@
...
+        return r;
+
+      if ((r = optdecl(expr)) != Skip)
+        return r;
+
+      if ((r = optnew(expr)) != Skip)
+        return r;
+
+      // We will have already found a tuple when trying to find a lambda, so
+      // this will always Skip.
```
Comment (mjp41) [inline src/parser/parser.cc:1200]: Add assert to check?

File: src/parser/parser.cc [hunk 2]
```diff
@@ -0,0 +1,2496 @@
...
+
+      if ((r = optspecialise(expr)) != Skip)
+        return r;
+
+      return Skip;
+    }
+
+    Result optpostorblock(Node<Expr>& expr)
+    {
+      // postfix <- atom ('.' (ident / symbol) / typeargs)*
```
Comment (mjp41) [inline src/parser/parser.cc:1331]: orblock in name, doesn't seem to be reflected in the comment?

**Review by mjp41 — APPROVED** [review]
LGTM.  There are some parts of the parser that are pretty tricky around postfix and deciding how to parse that could probably do with some documentation, before you forget how it works ;-)

File: src/parser/dnf.cc [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+    // (A | B) & (C | D) -> (A & C) | (A & D) | (B & C) | (B & D)
+    auto un = std::make_shared<UnionType>();
+    un->location = loc;
+
+    for (auto& ltype : left.types)
+    {
+      for (auto& rtype : right.types)
+      {
+        auto isect = std::make_shared<IsectType>();
+        isect->location = ltype->location;
```
[the remark this answers is not in the document]
Comment (sylvanc) [reply src/parser/dnf.cc:56]: I'll do a synthetic location that ranges over both ltype and rtype.

File: src/parser/dnf.cc [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+  Node<Type> isect_union(IsectType& left, UnionType& right, Location& loc)
+  {
+    // (A & B) & (C | D) -> (A & B & C) | (A & B & D)
+    auto un = std::make_shared<UnionType>();
+    un->location = loc;
+
+    for (auto& type : right.types)
+    {
+      auto isect = std::make_shared<IsectType>();
+      isect->location = type->location;
```
[the remark this answers is not in the document]
Comment (sylvanc) [reply src/parser/dnf.cc:36]: Another synthetic range!

File: src/parser/dnf.cc [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+
+  Node<Type> union_union(UnionType& left, UnionType& right, Location& loc)
+  {
+    // (A | B) & (C | D) -> (A & C) | (A & D) | (B & C) | (B & D)
+    auto un = std::make_shared<UnionType>();
+    un->location = loc;
+
+    for (auto& ltype : left.types)
+    {
+      for (auto& rtype : right.types)
```
[the remark this answers is not in the document]
Comment (sylvanc) [reply src/parser/dnf.cc:53]: Done.

File: src/parser/parser.cc [hunk 6]
```diff
@@ -0,0 +1,2496 @@
...
+      {
+        Node<Type> elem;
+
+        if (optisecttype(elem) != Success)
+        {
+          error() << loc() << "Expected a type" << line();
+          r = Error;
+        }
+
+        if (elem->kind() == Kind::UnionType)
```
[the remark this answers is not in the document]
Comment (sylvanc) [reply src/parser/parser.cc:1834]: Good call. Moved this into dnf.cc, giving optuniontype the same structure as optisecttype.

File: src/parser/escaping.h [hunk 7]
```diff
@@ -0,0 +1,18 @@
+// Copyright Microsoft and Project Verona Contributors.
+// SPDX-License-Identifier: MIT
+#pragma once
+
+#include <string_view>
+
+namespace verona::parser
```
[the remark this answers is not in the document]
Comment (sylvanc) [reply src/parser/escaping.h:7]: Comments added.

File: src/parser/ident.h [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+      store = std::make_shared<SourceDef>();
+    }
+
+    Location operator()(const char* text = "")
+    {
+      auto len = std::strlen(text);
+
+      if (len == 0)
+      {
+        auto h = "$" + std::to_string(hygienic++);
```
[the remark this answers is not in the document]
Comment (sylvanc) [reply src/parser/ident.h:27]: Yup! Hopefully not bad-clever.

File: src/parser/source.h [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+  {
+    std::string origin;
+    std::string contents;
+  };
+
+  using Source = std::shared_ptr<SourceDef>;
+
+  struct Location
+  {
+    Source source;
```
[the remark this answers is not in the document]
Comment (sylvanc) [reply src/parser/source.h:22]: I tried. Too many compiler errors to deal with, involving things like no const copy constructor for std::shared_ptr.

File: src/parser/fields.h [hunk 10]
```diff
@@ -0,0 +1,238 @@
+// Copyright Microsoft and Project Verona Contributors.
+// SPDX-License-Identifier: MIT
+#pragma once
+
+#include "ast.h"
+
```
[the remark this answers is not in the document]
Comment (sylvanc) [reply src/parser/fields.h:6]: Done.

File: src/parser/parser.cc [hunk 11]
```diff
@@ -0,0 +1,2496 @@
...
+
+      Result r = Success;
+      auto match = std::make_shared<Match>();
+      auto st = push(match);
+      match->location = previous.location;
+      expr = match;
+
+      if (opttuple(match->cond) != Success)
+      {
+        error() << loc() << "Expected a match condition" << line();
```
[the remark this answers is not in the document]
Comment (sylvanc) [reply src/parser/parser.cc:548]: Changed to "test-expression" to match F#.

File: src/parser/ast.h [hunk 12]
```diff
@@ -0,0 +1,713 @@
...
+
+    Kind kind()
+    {
+      return Kind::Case;
+    }
+  };
+
+  struct Match : BlockExpr
+  {
+    Node<Expr> cond;
```
[the remark this answers is not in the document]
Comment (sylvanc) [reply src/parser/ast.h:300]: Changed to `test` to match F#.

File: src/parser/parser.cc (hunk 1 above)
Comment (sylvanc) [reply src/parser/parser.cc:1200]: Done.

File: src/parser/parser.cc (hunk 2 above)
Comment (sylvanc) [reply src/parser/parser.cc:1331]: Improved comment!

**Comment by sylvanc** [comment]
The use of `wf << ast` is internal to the passes. It uses the infrastructure in `fields.h` and `dispatch.h` to forward the `<<` operator to the fields of AST nodes, recursively. The API externally for the dnf wellformedness pass can be seen on main.cc:37, and it looks like `dnf::wellformed(ast)`, returning a `bool`.

**Pull request merged**